examples/optimistic-updates-typescript/pages/index.tsx

```diff
@@ -0,0 +1,120 @@
+  return (
+    <QueryClientProvider client={client}>
+      <Example />
+    </QueryClientProvider>
+  )
+}
+
+type Todo = {
+  id: string
+  text: string
+}
+
+type TodoData = {
+  items: readonly Todo[]
+  ts: number
+}
+
+async function fetchTodos(): Promise<TodoData> {
+  const res = await axios.get('/api/data')
+  return res.data
+}
+
+function Example() {
+  const queryClient = useQueryClient()
+  const [text, setText] = React.useState('')
+  const { isFetching, ...queryInfo } = useQuery('todos', fetchTodos)
+
+  const addTodoMutation = useMutation(
+    (newTodo: string) => axios.post('/api/data', { text: newTodo }),
+    {
+      // Optimistically update the cache value on mutate, but store
+      // the old value and return it so that it's accessible in case of
+      // an error
+      onMutate: async newTodo => {
+        setText('')
+        await queryClient.cancelQueries('todos')
+
+        const previousValue = queryClient.getQueryData<TodoData>('todos')
+
+        if (previousValue) {
+          queryClient.setQueryData<TodoData>('todos', {
+            ...previousValue,
+            items: [
+              ...previousValue.items,
+              { id: Math.random().toString(), text: newTodo },
+            ],
+          })
+        }
+
+        return previousValue
+      },
+      // On failure, roll back to the previous value
+      onError: (err, variables, previousValue) => {
+        queryClient.setQueryData('todos', previousValue)
+      },
+      // // After success or failure, refetch the todos query
+      onSettled: () => {
+        queryClient.invalidateQueries('todos')
+      },
+    }
+  )
+
+  return (
+    <div>
+      <p>
+        In this example, new items can be created using a mutation. The new item
+        will be optimistically added to the list in hopes that the server
+        accepts the item. If it does, the list is refetched with the true items
+        from the list. Every now and then, the mutation may fail though. When
+        that happens, the previous list of items is restored and the list is
+        again refetched from the server.
+      </p>
+      <form
+        onSubmit={e => {
+          e.preventDefault()
+          addTodoMutation.mutate(text)
+        }}
+      >
+        <input
+          type="text"
+          onChange={event => setText(event.target.value)}
+          value={text}
+        />
+        <button disabled={addTodoMutation.isLoading}>Create</button>
+      </form>
+      <br />
+      {queryInfo.isSuccess && (
+        <>
+          <div>
+            Updated At: {new Date(queryInfo.data.ts).toLocaleTimeString()}
+          </div>
+          <ul>
+            {queryInfo.data.items.map(todo => (
+              <li key={todo.id}>{todo.text}</li>
+            ))}
+          </ul>
+          {isFetching && <div>Updating in background...</div>}
+        </>
+      )}
+      {queryInfo.isLoading && 'Loading'}
+      {queryInfo.error instanceof Error && queryInfo.error.message}
```

Comment:
do we want to advocate this behaviour (instanceOf check at runtime?) Otherwise, I would have to add generics to the `useQuery` call, which I think is not ideal either ....

Reply:
ah, it actually mentions this in the docs:

https://react-query-beta.tanstack.com/typescript#defining-custom-hooks

so I can change that if you want. Also, maybe we should amend that section a bit? If you use the `select` option, you have to actually define 3 generics.

Reply:
I made up my mind and converted this to a custom hook with a second usage that uses the `select` option to only display the length of todos in a counter, which only renders if the length of the array changes. Now I need the generics anyways 😅 . I hope its not too complicated now 🙃